### ISSUE TITLE
Automatic creation of string hashes

### ISSUE THREAD

**Issue by Tisten** [wc-duck/datalibrary#148]
Being able to write real strings in json and just store the string hash in the binary file (and putting backwards lookup info in a separate structure) and then being able to get strings again when converting back is an awesome space saver and debug feature which I'm missing in dl.

**Comment by wc-duck** [comment]
Im a bit hesitant to adding to much machinery around hashes directly in the core library mostly as I don't feel that DL can decide what hash-function to use and how a reverse-lookup should be performed. 

However, what could be done is adding 2 callbacks on the type lib for hashing and "unhashing" string-data and let the user implement them how they see fit.

This could be default to a hash-function using the internal dl_hash and not implementing a reverse-lookup.
When unpacking to text and there is no reverse-lookup the output value would be the hash as an int. As long as txt-pack accept int or string for a hash32/hash64 it should be an OK compromise!?

Any thoughts on that?

**Comment by wc-duck** [comment]
When thinking a bit möte about this one more thing to take into consideration popped up.

I know of another similar library that store the typelib after each packed instance for simpler decompression. I'm not really a fan of this as it eats lots of space that is mostly the same among packed days ( there can be up to 90% waste in many files!).
It is however convenient! 

Where am I going with this you might ask! Would it be worth adding an api or similar for users to alloc user data sections aster the packed instance where they could store things such as hashed strings and/or typelibs if they wish to?

**Comment by wc-duck** [comment]
Yes... user provided transforms is a good start. I however need to think about the api, especially if you want to switch types str -> uint32 for example.

**Comment by Tisten** [comment]
If these transforms are meant to be between json and binary, DL doesn't
have to care about the json format. So this could be between `const char*`
and the binary type. I.e the user needs to handle the float parsing and
writing of the "Euler" format and and just need to store the quaterion as 4
floats.

**Comment by wc-duck** [comment]
I see bo real problem with this as long as your storage in dl is pod... but if it starts to contain subdata such as strings and arrays it get way more complex! But maybe it is a valid restriction to only support conversion to pod-types?

**Comment by wc-duck** [comment]
Supporting conversion to a struct with arrays or pointers would add quite a bit of complexity... especially since then you could actually chain them as well!

**Comment by Tisten** [comment]
I agree. Starting out sith POD only will cover the cases mentioned so far. I might be optimistic when thinking about pointers and arrays, but I think they ought to work as well if need them. But leave them for later

**Comment by wc-duck** [comment]
An idea that popped up during a walk... this might be combined with "type-conversion".
Adding a function to convert from type-a to type-b could be used to upgrade data, but with a callback to that process you could also implement these kind of workshops. By specifying an "edit"-type and a "runtime" type and do a "conversion" in your buildpipe maybe 2 issues can be solved at once?

Something like that could possibly be used for "more complex" type-upgrades as well?

**Comment by Tisten** [comment]
Yes, I really like the concept of "data migration" which a "data upgrade" callback was called in the machinery which I worked with lately. They really made it possible to make advanced schema upgrades without breaking old data. They weren't bidirectional though, since a data upgrade isn't normally intended to solve the case of old code with data which is "too new".

I like the idea of having a single concept which a developer can utilize to solves both of these kinds of issues. But with that ambition it would need to support non-POD types. And it wouldn't be enough to do bidirectional conversion between a general "json string" and a specific type.

**Comment by wc-duck** [comment]
When I talk about data-conversion I don't think about json even being in the picture, I see something in the vain of 

dl_convert(packed_instance, &out_instance, new_type_id);

That would never need to go via json at all... not really sure how it would work, but if you could get something in your callback as:
error_code my_callback(old_stored_type* old_instance, new_stored_type* new_instance)
{
}

you could just write plain c to do you conversion? Especially together with the proposed "versioning of type" in another issue we could even generate the c-structs in a header for all versions of a type if applicable?

There is the question of how to implement your "string" -> hash, but if you have a string_hash type that is only a string-member and your hash32 type that is both types, it would work out quite nicely? I.e. your conversion function wwould get stored_type as "string_type*" and your new_stored_type would be a "hash32*".

that would make your conversion function to be:
new_type->hash = hash_the_string(old_type->str);

**Comment by Tisten** [comment]
Yes, I agree with your thoughts here. My comment about json was just to be clear about that the "old type" and "new type" is not about "old and new" but about "txt and binary" in that case, so the question about when to run the migration callback becomes more complex than to check for the highest version number since it needs to be bidirectional to and from the txt version.

I think I know you well enough to say that your callback signature is simplified, and that it should include a `void* user_data` which is registered with the callback function :)

**Comment by Tisten** [comment]
I took this prototype a bit further in https://github.com/Tisten/datalibrary/commits/patch_impl/ but never never finished it. The reason is that we so far don't use datalibrary for source data. We always create the datalibrary data from another source format and thus have not needed data migration. I like the idea of supporting data migration, but I don't really need it.

But lately I've thought more about the "human readable" part of datalibrary, and one feature which I would like to have is better visualization of key-value containers. I already have an implementation which can be stored in the binary version of datalibrary, and of cause it can be converted to json. But by doing so it just shows the implementation internals and the json becomes very hard to read for a human.

One solution could be to add a map/dictionary container into datalibrary. But I don't like that. It would need to define the data layout of the container, and questions like "should the data be sorted" are also locked down. The only feature I really need from datalibrary is the same as I need for the string hashes, and that is two callback functions for a specific typeid, one "binary to json" and one "json to binary".
